pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,600 @@
+		return
+	}
+
+	// We don't produce configs without root mapped and some runtimes assume it is mapped.
+	// Validate the file has something we produced and can digest.
+	if userNs.UIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: UID 0 must be mapped")
+		return
+	}
+
+	if userNs.GIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: GID 0 must be mapped")
+		return
+	}
+
+	hostId := userNs.UIDMappings[0].HostId
+	length := userNs.UIDMappings[0].Length
+
+	err = m.record(pod, hostId, length)
+	return
+}
+
+func (m *usernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err error) {
+	firstID, length, err := m.allocateOne(string(pod.UID))
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			m.Release(pod.UID)
+		}
+	}()
+
+	userNs = userNamespace{
+		UIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+		GIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+	}
+
+	return userNs, m.writeMappingsToFile(pod.UID, userNs)
+}
+
+// GetUserNamespaceMappings returns the configuration for the sandbox user namespace
+func (m *usernsManager) GetUserNamespaceMappings(pod *v1.Pod) (*runtimeapi.UserNamespace, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return nil, nil
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	if pod.Spec.HostUsers == nil || *pod.Spec.HostUsers == true {
+		return &runtimeapi.UserNamespace{
+			Mode: runtimeapi.NamespaceMode_NODE,
+		}, nil
+	}
+
+	content, err := m.readMappingsFromFile(pod.UID)
+	if err != nil && err != utilstore.ErrKeyNotFound {
+		return nil, err
+	}
+
+	var userNs userNamespace
+	if string(content) != "" {
+		userNs, err = m.parseUserNsFileAndRecord(pod.UID, content)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		userNs, err = m.createUserNs(pod)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var uids []*runtimeapi.IDMapping
+	var gids []*runtimeapi.IDMapping
+
+	for _, u := range userNs.UIDMappings {
+		uids = append(uids, &runtimeapi.IDMapping{
+			HostId:      u.HostId,
+			ContainerId: u.ContainerId,
+			Length:      u.Length,
+		})
+	}
+	for _, g := range userNs.GIDMappings {
+		gids = append(gids, &runtimeapi.IDMapping{
+			HostId:      g.HostId,
+			ContainerId: g.ContainerId,
+			Length:      g.Length,
+		})
+	}
+
+	return &runtimeapi.UserNamespace{
+		Mode: runtimeapi.NamespaceMode_POD,
+		Uids: uids,
+		Gids: gids,
+	}, nil
+}
+
+// cleanupOrphanedPodUsernsAllocations reconciliates the state of user namespace
+// allocations with the pods actually running. It frees any user namespace
+// allocation for orphaned pods.
+func (m *usernsManager) CleanupOrphanedPodUsernsAllocations(pods []*v1.Pod, runningPods []*kubecontainer.Pod) error {
```

Comment:
it should clear any user mappings files on disk for orphaned pods.

Reply:
It was added as part of the Release/releaseWithLock functions, so this does that by calling release too :)